test/test_public_bindings.py

```diff
@@ -272,6 +272,663 @@ def test_no_new_bindings(self):
+            "torch.utils.data.datapipes.iter.utils",
+            "torch.utils.data.datapipes.map",
+            "torch.utils.data.datapipes.map.callable",
+            "torch.utils.data.datapipes.map.combinatorics",
+            "torch.utils.data.datapipes.map.combining",
+            "torch.utils.data.datapipes.map.grouping",
+            "torch.utils.data.datapipes.map.utils",
+            "torch.utils.data.datapipes.utils",
+            "torch.utils.data.datapipes.utils.common",
+            "torch.utils.data.datapipes.utils.decoder",
+            "torch.utils.data.dataset",
+            "torch.utils.data.distributed",
+            "torch.utils.data.gen_pyi",
+            "torch.utils.data.graph",
+            "torch.utils.data.graph_settings",
+            "torch.utils.data.sampler",
+            "torch.utils.dlpack",
+            "torch.utils.ffi",
+            "torch.utils.file_baton",
+            "torch.utils.hipify",
+            "torch.utils.hipify.constants",
+            "torch.utils.hipify.cuda_to_hip_mappings",
+            "torch.utils.hipify.hipify_python",
+            "torch.utils.hipify.version",
+            "torch.utils.hooks",
+            "torch.utils.mkldnn",
+            "torch.utils.mobile_optimizer",
+            "torch.utils.model_dump",
+            "torch.utils.model_zoo",
+            "torch.utils.show_pickle",
+            "torch.utils.tensorboard",
+            "torch.utils.tensorboard.summary",
+            "torch.utils.tensorboard.writer",
+            "torch.utils.throughput_benchmark",
+            "torch.version"
+        }
+        for _, modname, ispkg in pkgutil.walk_packages(path=torch.__path__, prefix=torch.__name__ + '.'):
+            is_a_private_module = False
+            split_strs = modname.split('.')
+            for elem in split_strs:
+                if elem.startswith("_"):
+                    is_a_private_module = True
+                    break
+
+            if is_a_private_module:
+                continue
+
+            def module_starts_with_modname(elem, modname):
+                private_elem = elem.startswith('_')
+                obj = getattr(modname, elem)
+                elem_module = getattr(obj, '__module__') if hasattr(obj, '__module__') else None
+                elem_modname_starts_with_modname = elem_module is None or elem_module.startswith(modname)
+                if private_elem:
+                    assert elem_module is None or not elem_modname_starts_with_modname
+                else:
+                    assert elem_modname_starts_with_modname
+
+            if hasattr(modname, '__all__'):
+                assert not is_a_private_module
+                public_api = getattr(modname, '__all__')
+                all_api = dir(modname)
+                for elem in all_api:
+                    assert not (elem in public_api ^ module_starts_with_modname(elem, modname))
+            else:
+                assert modname in allowlist_modules_without_all
```

Comment:
Not sure why this is needed?
Both the version with all and the one with everything are ok. So I don't think we need an allow list to differentiate the two.

Reply:
I thought we wanted all public modules to have `__all__`?

Reply:
No you have a choice:
- Option 1 (the preferred one) that is using `__all__`
- Option 2 that uses `dir(mod)`

Both are ok.